python/pyarrow/tests/test_parquet.py

```diff
@@ -2936,7 +2939,34 @@ def test_write_to_dataset_with_partitions_and_index_name(
 @pytest.mark.pandas
 @parametrize_legacy_dataset
 def test_write_to_dataset_no_partitions(tempdir, use_legacy_dataset):
-    _test_write_to_dataset_no_partitions(str(tempdir))
+    _test_write_to_dataset_no_partitions(str(tempdir), use_legacy_dataset)
+
+
+@pytest.mark.pandas
+@parametrize_legacy_dataset
+def test_write_to_dataset_pathlib(tempdir, use_legacy_dataset):
+    _test_write_to_dataset_with_partitions(
+        tempdir / "test1", use_legacy_dataset)
+    _test_write_to_dataset_no_partitions(
+        tempdir / "test2", use_legacy_dataset)
+
+
+@pytest.mark.pandas
+@pytest.mark.s3
+@parametrize_legacy_dataset
+def test_write_to_dataset_pathlib_nonlocal(
+    tempdir, s3_example_s3fs, use_legacy_dataset
+):
+    # pathlib paths are only accepted for local files
+    fs, _ = s3_example_s3fs
```

Comment:
Hmm. Ideally we don't require S3 for such tests (because it's optional, and also because it makes tests much more expensive to run). Can we use a mock filesystem or something?

Reply:
The problem is that this is using the "legacy" filesystems (I am going to add support for the new filesystems when the dataset writing with parquet is merged, and even then it might take a different code path), and I don't think we have a mock filesystem in pyarrow.filesytem? (only the local, since we also can't use hdfs)

Reply:
Ah, I see. This is good enough then.

Reply:
I could of course use a "mock" (eg in-memory) filesytem from fsspec. Then it still uses an optional dependency in this test, but at least is not that slow as S3